File: scripts/shared/github-output.mjs
#!/usr/bin/env node

import fs from "node:fs";
import crypto from "node:crypto";

function buildDelimiter(serialized) {
  const base = `EOF_${crypto.randomUUID().replace(/-/g, "")}`;
  if (!serialized.includes(base)) return base;

  let suffix = 0;
  while (serialized.includes(`${base}_${suffix}`)) {
    suffix += 1;
  }

  return `${base}_${suffix}`;
}

export function writeGithubOutput(name, value, outputPath = process.env.GITHUB_OUTPUT) {
  if (!outputPath) return;

  const serialized = String(value ?? "");
  const delimiter = buildDelimiter(serialized);
  fs.appendFileSync(outputPath, `${name}<<${delimiter}\n${serialized}\n${delimiter}\n`);
}

File: scripts/shared/github_output.py
from __future__ import annotations

import secrets
from pathlib import Path

from .types import JSONScalar


def _build_delimiter(serialized: str) -> str:
    base = f"EOF_{secrets.token_hex(16)}"
    if base not in serialized:
        return base

    suffix = 0
    while f"{base}_{suffix}" in serialized:
        suffix += 1

    return f"{base}_{suffix}"


def write_github_output(name: str, value: JSONScalar, output_path: str | None) -> None:
    if not output_path:
        return

    serialized = str("" if value is None else value)
    delimiter = _build_delimiter(serialized)
    payload = f"{name}<<{delimiter}\n{serialized}\n{delimiter}\n"

    path = Path(output_path)
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("a", encoding="utf-8") as handle:
        handle.write(payload)
